sys/decl.go

```diff
@@ -110,22 +111,17 @@ const (
 
 type BufferType struct {
 	TypeCommon
-	Kind BufferKind
+	Kind       BufferKind
+	RangeBegin uintptr // for BufferBlobRange kind
+	RangeEnd   uintptr // for BufferBlobRange kind
 }
 
 func (t BufferType) Size() uintptr {
-	switch t.Kind {
-	case BufferAlgType:
-		return 14
-	case BufferAlgName:
-		return 64
-	default:
-		panic(fmt.Sprintf("buffer size is not statically known: %v", t.Name()))
-	}
+	panic("not called")
```

Comment:
There is something wrong, this must be called.

The problem is that after recent changes we add struct alignment before struct fields are initialized. So effectively all structs are empty addAlignment. This happens because order of init function execution is unspecified. With the following patch these panics fire. The bad news are that StructType panics also fire.

```
diff --git a/sys/decl.go b/sys/decl.go
index 4e54111..9710283 100644
--- a/sys/decl.go
+++ b/sys/decl.go
@@ -498,6 +498,7 @@ var (
 )
 
 func init() {
+       initStructFields()
        initCalls()
        initResources()
        initAlign()
diff --git a/sysgen/sysgen.go b/sysgen/sysgen.go
index 1a84a53..4a943c5 100644
--- a/sysgen/sysgen.go
+++ b/sysgen/sysgen.go
@@ -283,7 +283,7 @@ func generateStructs(desc *Description, consts map[string]uint64, out io.Writer)
        }
        fmt.Fprintf(out, "}\n")
 
-       fmt.Fprintf(out, "func init() {\n")
+       fmt.Fprintf(out, "func initStructFields() {\n")
        for _, str := range structArray {
                typ := "StructType"
                fields := "Fields"
```


We need to step back, figure out the testing story (filed https://github.com/google/syzkaller/issues/78 for this), revert removal of struct Size/Align, then proceed with this change.